### ISSUE TITLE
Meta issue: No new tests until we have cleaned up the existing test infrastructure

### ISSUE THREAD

**Issue by heartsucker** [freedomofpress/securedrop#2877]
# Feature request

## Description

We omit many best practices for our tests. One thing we should do is properly use `pytest`'s fixtures to generate our apps. This can be done with:

```python
@pytest.fixture(scope='session')
def source_app(config):
    app = source_app.create_app(config)
    app.testing = True
    app.config['WTF_CSRF_ENABLED'] = False
    app.config['SERVER_NAME'] = 'localhost'

    with app.app_context():
        db.create_all()
        with app.test_client() as app:
            yield app


@pytest.fixture(scope='session')
def journalist_app(config):
    app = journalist_app.create_app(config)
    app.testing = True
    app.config['WTF_CSRF_ENABLED'] = False
    app.config['SERVER_NAME'] = 'localhost'

    with app.app_context():
        db.create_all()
        with app.test_client() as app:
            yield app
```

Right now we have issues (#1444) that point out some of the errors we hit. Are are also manually mucking around with the `g` and object for some of our tests which is extremely brittle. The best thing to do to get a user/admin to a logged in state is to `POST` to `/login`.

In short, we really should work with fixtures and be mindful of the applications and context we're using since introducing #2459 (Flask-SQLAlchemy) will make it so that we have to pay attention to these contexts much more.

## User Stories

As a developer, I find that the hardest part of adding new features and refactoring is always, always in the tests. This is problematic since editing a test in parallel with the app code is how bugs get introduced.

**Comment by heartsucker** [comment]
I also have written a test suite that boots a new container for the DB for each test so that every test case starts with a clean slate. This allows for parallel tests too since right now we have some stuff like `max[user.id for user in Journalist.query.all()])` which won't work in parallel land.

For now, since we're not on Postgres yet, we could just use a new database file that gets init'd at the `scope='function'` granularity. This would have minimal overhead.

**Comment by heartsucker** [comment]
Another thing I've found is that tests sometimes have implicit requirements on the state of other tests. I reordered some and things broke. Making the fixtures all have the `function` scope forces a test to demonstrate every step

```python
def test_foo(self, journalist_app):
    j = self._create_new_journalist()
    self._login_journalist(jouranlist_app, j)
    resp = journalist_app.post(url_for('foo.endpoint', data=dict(bar='baz')))
    assert 'quux' in resp.data.decode('utf-8')
```

**Comment by conorsch** [comment]
Great discussion here, @heartsucker. Allow me to clarify that "no new tests" refers specifically to the application code tests. Within that category, you seem to mean unit tests specifically, but perhaps also functional tests (i.e. Selenium). 

According to my read of your posts, you do _not_ expect configuration tests to be omitted from such refactoring. While I certainly have my misgivings about technical debt on the config test side (I'm looking at you, `testinfra/test.py`), folks can continue to submit PRs modifying non-application-code tests even before the improvements you propose land in the `develop` branch.

Please correct me if I've misunderstood your request!

**Comment by heartsucker** [comment]
Yep, app code only. Once we get #2895 merged, I'm going to try to tackle this before I do all the Alembic stuff. In fact, I'm going to say this is a blocker for #1419 and #2225 since those two are going to likely add significant changes, and if we don't do this first, I'm going to have a very sad time implementing those two.

**Comment by heartsucker** [comment]
Right now we use `flask_testing` which after the migration to `pytest` fixtures will only be necessary to provide us with some convenience methods like `self.assertMessageFlashed`. A lot of the other methods like `self.assert200` can be trivially replaced with `assert resp.status_code == 200` (which honestly seems clearer to me than just asserting that *something* is 200).

Since this will impact a lot of the tests and may require us to write some helper utilities of our own, I wanted to see if anyone has any particular objection to this.

**Comment by heartsucker** [comment]
In PR #2948, I implemented some of this logic. Y'all should look there and tell me what you think.

**Comment by heartsucker** [comment]
This is blocked by #2969.

**Comment by heartsucker** [comment]
Tracking progress:

- [ ] functional/test_admin_interface.py
- [ ] functional/test_journalist.py
- [ ] functional/test_source_notfound.py
- [ ] functional/test_source.py
- [ ] functional/test_source_session_timeout.py
- [ ] functional/test_source_warnings.py
- [ ] functional/test_submission_not_in_memory.py
- [ ] functional/test_submit_and_retrieve_file.py
- [ ] functional/test_submit_and_retrieve_message.py
- [ ] pages-layout/test_journalist.py
- [ ] pages-layout/test_source.py
- [ ] test_2fa.py
- [ ] test_crypto_util.py
- [ ] test_db.py
- [ ] test_i18n.py
- [ ] test_i18n_tool.py
- [ ] test_integration.py
- [ ] test_journalist.py
- [ ] test_manage.py
- [ ] test_secure_tempfile.py
- [x] test_source.py
- [ ] test_store.py
- [ ] test_template_filters.py

**Comment by heartsucker** [comment]
I'm marking this ticket as closed because all non-functional tests have been converted to pytest/fixtures. Converting the functional tests is a fundamentally different endeavor for a few reasons. I'm going to open a ticket for that.
